scrapy_poet/cache.py

```diff
@@ -20,54 +20,46 @@ def close(self) -> None:  # noqa: B027
-            path,
-            tablename=tablename,
-            autocommit=True,
-            encode=self.encode,
-            decode=self.decode,
-        )
-
-    def encode(self, obj: Any) -> memoryview:
-        # based on sqlitedict.encode
-        data = pickle.dumps(obj, pickle.HIGHEST_PROTOCOL)
-        if self.compressed:
-            data = gzip.compress(data, compresslevel=3)
-        return sqlite3.Binary(data)
+    def __init__(self, directory: Union[str, os.PathLike]) -> None:
+        self.directory = directory
 
-    def decode(self, obj: Any) -> Any:
-        # based on sqlitedict.decode
-        data = bytes(obj)
-        if self.compressed:
-            # gzip is slightly less efficient than raw zlib, but it does
-            # e.g. crc checks out of box
-            data = gzip.decompress(data)
-        return pickle.loads(data)
+    def __getitem__(self, fingerprint: str) -> SerializedData:
+        storage = SerializedDataFileStorage(self._get_directory_path(fingerprint))
+        try:
+            serialized_data = storage.read()
+        except FileNotFoundError:
+            raise KeyError(f"Fingerprint '{fingerprint}' not found in cache")
+        return serialized_data
 
-    def __str__(self) -> str:
-        return (  # pragma: no cover
-            f"SqlitedictCache <{self.db.filename} | "
-            f"compressed: {self.compressed} | "
-            f"{len(self.db)} records>"
-        )
+    def __setitem__(
+        self, fingerprint: str, value: Union[SerializedData, Exception]
+    ) -> None:
+        if isinstance(value, Exception):
+            self.write_exception(fingerprint, value)
+        else:
+            storage_path = Path(self._get_directory_path(fingerprint))
+            if not storage_path.exists():
+                storage_path.mkdir(parents=True)
+            storage = SerializedDataFileStorage(self._get_directory_path(fingerprint))
+            storage.write(value)
 
-    def __repr__(self) -> str:
-        return f"SqlitedictCache({self.path!r}, compressed={self.compressed})"  # pragma: no cover
+    def write_exception(self, fingerprint: str, exception: Exception) -> None:
+        with open(self._get_exception_file_path(fingerprint), "wb") as file:
+            pickle.dump(exception, file)
 
-    def __getitem__(self, fingerprint: str) -> Any:
-        return self.db[fingerprint]
+    def _get_directory_path(self, fingerprint: str) -> str:
+        return os.path.join(self.directory, fingerprint)
 
-    def __setitem__(self, fingerprint: str, value: Any) -> None:
-        self.db[fingerprint] = value
+    def _get_exception_file_path(self, fingerprint: str) -> str:
+        """Save exception inside self.directory, so that `storage.read()` can read it correctly"""
+        return os.path.join(
+            self._get_directory_path(fingerprint), f"{fingerprint}.error"
+        )
 
-    def close(self) -> None:
-        self.db.close()
+    # TODO: Add option for compressed cache
```

Comment:
@kmike @wRAR Do we want to implement this here or leave it for a future iteration?

Reply:
I think it should be done separately and I see it as some kind of pluggable `SerializedDataFileStorage` implementations in web-poet.